src/test_application/VehicleBuilder.cpp

```diff
@@ -80,17 +80,162 @@ partindex_t VehicleBuilder::add_part(
+    {
+        return {osp::nullvalue<mach_t>(), nullptr}; // no machines of type
+    }
+
+    for (size_t i = 0; i < m_vehicle.m_machines[id].size(); i ++)
+    {
+        osp::BlueprintMachine& machineBp = m_vehicle.m_machines[id][i];
+        if (machineBp.m_partIndex == uint32_t(part))
+        {
+            // Found
+            return {mach_t(machineBp.m_protoMachineIndex), &machineBp};
+        }
+    }
+    return {osp::nullvalue<mach_t>(), nullptr};
+}
+
+
+nodeindex_t<void> VehicleBuilder::wire_node_add(osp::wire_id_t id,
+                                           osp::NodeMap_t config)
+{
+    // Resize wire type vector if it's too small
+    if (m_vehicle.m_wireNodes.size() <= id)
+    {
+        m_vehicle.m_wireNodes.resize(size_t(id) + 1);
+    }
+
+    // Get vector of nodes
+    std::vector<osp::BlueprintWireNode> &rNodes
+            = m_vehicle.m_wireNodes[size_t(id)];
+    uint32_t index = rNodes.size();
+
+    // Create new node
+    rNodes.emplace_back(osp::BlueprintWireNode{config, {}});
+    return nodeindex_t<void>(index);
+}
+
+linkindex_t<void> VehicleBuilder::wire_connect(
+        osp::wire_id_t id, nodeindex_t<void> node, osp::NodeMap_t config,
+        part_t part, mach_t mach, portindex_t<void> port)
+{
+    if (m_vehicle.m_wireNodes.size() <= id)
+    {
+        // No node of type is even registered
+        return osp::nullvalue< linkindex_t<void> >();
+    }
+
+    // Keep track of wire connections
+    auto nodeIt = m_nodeIndexMap.try_emplace({id, part, mach, port}, node);
+    if (!nodeIt.second)
+    {
+        // Error! Port already connected!
+        return osp::nullvalue<linkindex_t<void>>();
+    }
+
+    // Get the BlueprintWireNode
+    std::vector<osp::BlueprintWireNode> &rNodes
+            = m_vehicle.m_wireNodes[size_t(id)];
+    osp::BlueprintWireNode &rNode = rNodes[size_t(node)];
+
+    // Create BlueprintWireLink
+    uint32_t index = rNode.m_links.size();
+    rNode.m_links.emplace_back(
+            osp::BlueprintWireLink{config, uint32_t(part), uint16_t(mach),
+                                   uint16_t(port)});
+
+    // Update panel. There's no easy way to find the correct BlueprintWirePanel
+    // from m_vehicle.m_wirePanels from a wire index, part index, and machine
+    // index. We can search the vector if the right panel already exists, but
+    // here we're using a map
+    uint32_t newIndex = m_vehicle.m_wirePanels.size();
+    auto panelIt = m_panelIndexMap.try_emplace({id, part, mach}, newIndex);
+
+    uint16_t minPortCount = uint16_t(port) + 1;
+
+    if (panelIt.second)
+    {
+        // new panel entry added
+        m_vehicle.m_wirePanels.resize(std::max(m_vehicle.m_wirePanels.size(),
+                                               size_t(id + 1)));
+        m_vehicle.m_wirePanels[size_t(id)].emplace_back(
+                    osp::BlueprintWirePanel{uint32_t(part), uint16_t(mach),
+                                            uint16_t(minPortCount)});
```

Comment:
why do you have so much casting going on here? minPortCount is already a uint16_t